pysmt/cmd/installers.py

```diff
@@ -0,0 +1,630 @@
+import re
+import sys
+import os
+import tarfile
+import zipfile
+import pip
+import subprocess
+import platform
+import shutil
+
+from contextlib import contextmanager
+
+from six.moves import xrange
+from six.moves.urllib import request as urllib2
+from six.moves.urllib.error import HTTPError
+
+
+def untar(fname, directory, mode='r:gz'):
+    """Extracts the tarfile using the specified mode in the given directory."""
+    tfile = tarfile.open(fname, mode)
+    tfile.extractall(directory)
+
+def unzip(fname, directory):
+    """Unzips the given archive into the given directory"""
+    myzip = zipfile.ZipFile(fname, "r")
+    myzip.extractall(directory)
+    myzip.close()
+
+@contextmanager
+def TemporaryPath(path):
+    """Context that substitutes the system path to test for API presence or absence"""
+    old_path = list(sys.path)
+    try:
+        sys.path = path + sys.path
+        yield
+    finally:
+        sys.path = old_path
+
+
+
+def get_architecture_bits():
+    """Returns the native word width of this architecture. E.g. 32 or 64"""
+    is_64bits = sys.maxsize > 2**32
+    if is_64bits:
+        return 64
+    else:
+        return 32
+
+
+def get_python_version():
+    """Returns the current python version as string E.g. '2.7'"""
+    return "%d.%d" % sys.version_info[0:2]
+
+
+def get_os():
+    """Returns the current OS name E.g. 'linux'"""
+    return platform.system().lower()
+
+
+def get_architecture():
+    """Returns the short name of the architecture in use. E.g. 'x86_64'"""
+    return platform.machine()
+
+
+
+class SolverInstaller(object):
+
+    SOLVER = None
+
+    def __init__(self, install_dir, bindings_dir, solver_version, os_name=None,
+                 architecture=None, python_version=None, mirror_link=None):
+        self.bindings_dir = bindings_dir
+        self.install_dir = install_dir
+        self.solver_version = solver_version
+        self.os_name = os_name if os_name else get_os()
+        self.architecture = architecture if architecture else get_architecture()
+        self.python_version = python_version if python_version else get_python_version()
+        self.mirror_link = mirror_link
+
+        self.trials_404 = 3
+        self.base_dir = None
+
+        self.download_links = []
+        self.archive_path = None
+
+        self.base_dir = os.path.join(self.install_dir, self.SOLVER)
+        if not os.path.exists(self.base_dir):
+            os.mkdir(self.base_dir)
+
+
+    def download(self):
+        """Downloads the archive from one of the mirrors"""
+        if not os.path.exists(self.archive_path):
+            for _ in xrange(self.trials_404):
+                for link in self.download_links:
```

Comment:
For debugging purposes, I think it would be useful to know from which mirror the package was downloaded. Maybe the index is enough...